src/hardware/pcspeaker_discrete.cpp

```diff
@@ -0,0 +1,491 @@
+		break;
+	default:
+#if C_DEBUG
+		LOG_WARNING("PCSPEAKER: Unhandled speaker PIT mode: %s", PitModeToString(pit_mode));
+#endif
+		return;
+	}
+	// Activate the channel after queuing new speaker entries
+	channel->Enable(true);
+}
+
+// Returns the amp_neutral voltage if the speaker's  fully faded,
+// otherwise returns the fallback if the speaker is active.
+double PcSpeakerDiscrete::NeutralOr(double fallback)
+{
+	return !idle_countdown ? amp_neutral : fallback;
+}
+
+// Returns, in order of preference:
+// - Neutral voltage, if the speaker's fully faded
+// - The last active PIT voltage to stitch on-going playback
+// - The fallback voltage to kick start a new sound pattern
+double PcSpeakerDiscrete::NeutralLastPitOr(double fallback)
+{
+	const bool use_last = std::isgreater(fabs(pit_last), amp_neutral);
+	return NeutralOr(use_last ? pit_last : fallback);
+}
+
+// PWM-mode activation
+void PcSpeakerDiscrete::SetType(const PpiPortB &b)
+{
+	const auto newindex = PIC_TickIndex();
+	ForwardPIT(newindex);
+
+	prev_port_b_state.data = port_b_state.data;
+
+	switch (b.timerGateAndSpeakerOutput) {
+	case 0b00:
+		port_b_state.timerGateOutput = false;
+		port_b_state.speakerOutput   = false;
+		AddDelayEntry(newindex, NeutralOr(amp_negative));
+		break;
+	case 0b01:
+		port_b_state.timerGateOutput = false;
+		port_b_state.speakerOutput   = true;
+		AddDelayEntry(newindex, NeutralOr(amp_negative));
+		break;
+	case 0b10:
+		port_b_state.timerGateOutput = true;
+		port_b_state.speakerOutput   = false;
+		AddDelayEntry(newindex, NeutralOr(amp_positive));
+		break;
+	case 0b11:
+		port_b_state.timerGateOutput = true;
+		port_b_state.speakerOutput   = true;
+		AddDelayEntry(newindex, NeutralLastPitOr(amp_positive));
+		break;
+	};
+
+	// Activate the channel after queuing new speaker entries
+	channel->Enable(true);
+}
+
+// After the speaker has idled (as defined when the number of speaker
+// movements is zero) for a short time, wind down the DC-offset, and
+// then halt the channel.
+void PcSpeakerDiscrete::PlayOrFadeout(const uint16_t speaker_movements,
+                                      size_t requested_samples, int16_t *buffer)
+{
+	if (speaker_movements && requested_samples) {
+		static_assert(idle_grace_time_ms > 0, "Algorithm depends on a non-zero grace time");
+		idle_countdown = idle_grace_time_ms;
+		dc_silencer.Reset();
+	} else if (idle_countdown > 0) {
+		idle_countdown--;
+		last_played_sample = buffer[requested_samples - 1];
+	} else if (!dc_silencer.Generate(last_played_sample, requested_samples, buffer)) {
+		channel->Enable(false);
+	}
+
+	channel->AddSamples_m16(check_cast<uint16_t>(requested_samples), buffer);
+}
+
+void PcSpeakerDiscrete::ChannelCallback(uint16_t frames)
+{
+	constexpr uint16_t render_frames = 64;
+	int16_t buf[render_frames];
+
+	ForwardPIT(1);
+	last_index            = 0.0;
+	uint16_t pos          = 0u;
+	auto sample_base      = 0.0;
+	const auto sample_add = (1.0001) / frames;
```

Comment:
Why 1.0001 and not just 1.0?

Reply:
Thinking about it more, I think it represents the fraction of the current 1ms tick per sample. Still not positive about the tiny bit extra, though.

Reply:
Edit - Comment from the future: the change we accepted in this PR caused this regression: https://github.com/dosbox-staging/dosbox-staging/issues/1776

The reason for the small extra time factor is now better understood and described in the code: https://github.com/dosbox-staging/dosbox-staging/pull/1778/commits/2377e36fae25acd0dd2f29d30f556c56729d5ce3